Bluetooth: GATT: Allow sending less bytes than requested

Vudentz opened the pull request.

Vudentz committed:
Bluetooth: ATT: Add bt_att_get_payload_len

This adds bt_att_get_payload_len function which can be used to read the
available space for the payload given an opcode.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Allow sending less bytes than requested

This makes bt_gatt_{write_without_rsp, notify, indicate} to send less
bytes than requested like most IO procedures since this is a lot more
convenient than having to calculate the allowed payload in the
application side and then adjust the length it that doesn't fit.

If the application requires to send as packet it will need to call
bt_gatt_get_mtu and adjust it with bt_gatt_exchange_mtu in order before
it can use these procedures.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on subsys/bluetooth/host/att.c:
```diff
@@ -2285,3 +2285,24 @@ void bt_att_req_cancel(struct bt_conn *conn, struct bt_att_req *req)
 	att_req_destroy(req);
 }
+
+u16_t bt_att_get_payload(struct bt_conn *conn, uint8_t op)
```
I know the name gets rather long, but this should really be get_payload_len(), since it doesn't return the payload, rather its length.

Vudentz replied:
Sure, I will adjust that.

jhedberg commented:
For the record, there are protocols where it's critical that the requested amount gets sent or then nothing gets sent. The Mesh GATT Proxy protocol is one such example. However, I guess it's then the responsibility of the API user to check the MTU first.

Vudentz commented:
> For the record, there are protocols where it's critical that the requested amount gets sent or then nothing gets sent. The Mesh GATT Proxy protocol is one such example. However, I guess it's then the responsibility of the API user to check the MTU first.

Yep, the rationale is that these type of uses is much less frequent than write as much as possible usage, and it is also simpler to check it before hand and don't send anything.

Vudentz commented:
@carlescufi Any option on this? I mean it turns APIs such like bt_gatt_notify to work more as a stream rather than a packet, though perhaps it is better to introduce a flag to control this behavior since that might depend on the attribute.

Vudentz committed:
Bluetooth: GATT: Introduce procedure flags

This introduces procedure flags for bt_gatt_write_without_rsp,
bt_gatt_notify and bt_gatt_indicate so the user can select things like
signing a write and use of stream mode when necessary.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>
